datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -1958,6 +1964,90 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn row_group_predicate_between() -> Result<()> {
+        let schema = Schema::new(vec![
+            Field::new("c1", DataType::Int32, false),
+            Field::new("c2", DataType::Int32, false),
+        ]);
+
+        // test c1 BETWEEN 1 AND 5
+        let expr1 = Expr::Between(Between {
+            expr: Box::new(col("c1")),
+            negated: false,
+            low: Box::new(lit(1)),
+            high: Box::new(lit(5)),
+        });
+
+        // test 1 <= c1 <= 5
+        let expr2 = col("c1").gt_eq(lit(1)).and(col("c1").lt_eq(lit(5)));
+
+        let predicate_expr1 =
+            test_build_predicate_expression(&expr1, &schema, &mut RequiredColumns::new());
+
+        let predicate_expr2 =
+            test_build_predicate_expression(&expr2, &schema, &mut RequiredColumns::new());
+        assert_eq!(predicate_expr1.to_string(), predicate_expr2.to_string());
+
+        Ok(())
+    }
+
+    #[test]
+    fn row_group_predicate_between_with_in_list() -> Result<()> {
+        let schema = Schema::new(vec![
+            Field::new("c1", DataType::Int32, false),
+            Field::new("c2", DataType::Int32, false),
+        ]);
+        // test c1 in(1, 2)
+        let expr1 = Expr::InList(InList::new(
+            Box::new(col("c1")),
+            vec![lit(1), lit(2)],
+            false,
+        ));
+
+        // test c2 BETWEEN 4 AND 5
+        let expr2 = Expr::Between(Between {
+            expr: Box::new(col("c2")),
+            negated: false,
+            low: Box::new(lit(4)),
+            high: Box::new(lit(5)),
+        });
+
+        // test c1 in(1, 2) and c2 BETWEEN 4 AND 5
+        let expr3 = Expr::BinaryExpr(BinaryExpr {
+            left: Box::new(expr1),
+            op: Operator::And,
+            right: Box::new(expr2),
+        });
+
+        let expected_expr = "(c1_min@0 <= 1 AND 1 <= c1_max@1 OR c1_min@0 <= 2 AND 2 <= c1_max@1) AND c2_max@2 >= 4 AND c2_min@3 <= 5";
+        let predicate_expr =
+            test_build_predicate_expression(&expr3, &schema, &mut RequiredColumns::new());
+        assert_eq!(predicate_expr.to_string(), expected_expr);
+
+        Ok(())
+    }
+
+    #[test]
+    fn row_group_predicate_in_list_to_many_values() -> Result<()> {
+        let schema = Schema::new(vec![Field::new("c1", DataType::Int32, false)]);
+        // test c1 in(1..21)
+        // in pruning.rs has MAX_LIST_VALUE_SIZE_REWRITE = 20, more than this value will be rewrite
```

Comment:
👍

Reply:
It is good to document the current behavior in a test, but I feel like we can do better -- like https://github.com/apache/arrow-datafusion/issues/7869 says even if we can't prove the expression is not true due to a large inlist, we shouldn't just disable pruning entirely